src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,554 @@
+/// to the `cargo-test` step in `ci/test/pipeline.template.yml` and setting
+/// `MZ_S3_UPLOADER_TEST_S3_BUCKET` in
+/// `ci/test/cargo-test/mzcompose.py`.
+///
+/// For a Materialize developer, to opt in to these tests locally for
+/// development, follow the AWS access guide:
+///
+/// ```text
+/// https://www.notion.so/materialize/AWS-access-5fbd9513dcdc4e11a7591e8caa5f63fe
+/// ```
+///
+/// then running `source src/aws-util/src/setup_test_env_mz.sh`. You will also have
+/// to run `aws sso login` if you haven't recently.
+#[cfg(test)]
+mod tests {
+    use bytes::Bytes;
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::{defaults, s3};
+
+    fn s3_bucket_path_for_test() -> Option<(String, String)> {
+        let bucket = match std::env::var("MZ_S3_UPLOADER_TEST_S3_BUCKET") {
+            Ok(bucket) => bucket,
+            Err(_) => {
+                if mz_ore::env::is_var_truthy("CI") {
+                    panic!("CI is supposed to run this test but something has gone wrong!");
+                }
+                return None;
+            }
+        };
+
+        let prefix = Uuid::new_v4().to_string();
+        let path = format!("cargo_test/{}/file", prefix);
+        Some((bucket, path))
+    }
+
+    #[mz_ore::test(tokio::test(flavor = "multi_thread"))]
+    #[cfg_attr(coverage, ignore)] // https://github.com/MaterializeInc/materialize/issues/18898
+    #[cfg_attr(miri, ignore)] // error: unsupported operation: can't call foreign function `TLS_method` on OS `linux`
+    async fn multi_part_upload_success() -> Result<(), S3MultiPartUploadError> {
+        let sdk_config = defaults().load().await;
+        let (bucket, path) = match s3_bucket_path_for_test() {
+            Some(tuple) => tuple,
+            None => return Ok(()),
+        };
+
+        let config = S3MultiPartUploaderConfig::default();
+        let mut uploader =
+            S3MultiPartUploader::try_new(&sdk_config, bucket.clone(), path.clone(), config).await?;
+
+        let expected_data = "onetwothree";
+        let _ = uploader.add_chunk(Bytes::from("one")).await?;
+        let _ = uploader.add_chunk(Bytes::from("two")).await?;
+        let _ = uploader.add_chunk(Bytes::from("three")).await?;
+
+        // This should trigger one single part upload.
+        let CompletedUpload {
+            part_count,
+            total_bytes_uploaded,
+        } = uploader.finish().await?;
+
+        // Getting the uploaded object from s3 and validating the contents.
+        let s3_client = s3::new_client(&sdk_config);
+        let uploaded_object = s3_client
+            .get_object()
+            .bucket(bucket)
+            .key(path)
+            .part_number(1) // fetching a particular part, so that the `parts_count` is populated in the result
+            .send()
+            .await
+            .unwrap();
+
+        let uploaded_parts_count: u32 = uploaded_object.parts_count().unwrap().try_into().unwrap();
+        assert_eq!(uploaded_parts_count, part_count);
```

Comment:
Probably worth `assert_eq!(part_count, 1)`? To verify that everyone agrees that the part count is actually 1?